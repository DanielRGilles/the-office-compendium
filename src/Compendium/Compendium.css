.main-body {
    margin: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 100;
    background-color: cornflowerblue;
    height: 100;
}
.main-cnt {
    display: grid;
    grid-template-columns: 25% 1fr 25%;
}