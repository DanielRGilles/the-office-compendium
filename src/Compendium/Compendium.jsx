import {  useState, useEffect } from "react";
import QuoteList from "../Components/QuoteList/QuoteList";
import { fetchQuotes, filteredQuotes, fetchSearchQuery, randomQuotes } from "../Services/Quotes";
import './Compendium.css'
import Header from "../Components/Header";
import Controls from "../Components/Controls/Controls";

export default function Compendium() {
  const [loading, setLoading] = useState(true);
  const [quotes, setQuotes] = useState([]);
  const [characterName, setCharacterName] = useState('');
  const [query, setQuery ] = useState('');
  const [numOfRandom, setNum ] = useState();
    
  // on page load this hook will query the api for 40 quotes, then randomly grab 10 and pass those down to render
    useEffect(() => { 
        async function getQuotes() {
            const quoteList = await fetchQuotes();
            setQuotes(quoteList);
            setLoading(false);
        };
    getQuotes();
    }, [])
    
// this hook will listen for the character value to change and if so will fetch the max 40 quotes and render those that match the character name
    useEffect(() => {
        if (!characterName) return;
        async function getFiltered() {
            setLoading(true)
            if(characterName !== 'all') {    
                const quoteList = await filteredQuotes(characterName);
                setQuotes(quoteList);
                
            } else {
                const quoteList = await fetchQuotes();
                setQuotes(quoteList);
              } setLoading(false);

        };
        getFiltered();

        }, [characterName])
    
// this form will pass down the query string so that a user can search for all quotes by a character name OR search the max 40 quotes for specific words/letters and return those
  const handleSubmit = async (event) => {
    event.preventDefault();
    setLoading(true);
    const searched = await fetchSearchQuery(query)
      setQuotes(searched);
      setLoading(false);
      setQuery('');
      setCharacterName('');
  }
  // this event triggered on button click that will individually grab random quotes from the api
    const handleRandomSubmit =  async (event) => {
        event.preventDefault();
        setLoading(true); 
        const random = await randomQuotes(numOfRandom)
        setQuotes(random);
        setLoading(false);
         }


    return (
    <div className='main-body'>
        <Header/>
        <Controls 
        handleSubmit={handleSubmit}
        handleWordChange={setQuery}
        query={query}
        characterName={characterName} 
        setCharacterName={setCharacterName}
        numOfRandom={numOfRandom}
        setRandomQuotes={handleRandomSubmit}
        setNum={setNum}
       />
            {loading 
                ? (<h1 className='load-spin'>Loading...</h1>)
                : (
                    <main className='main-cnt'> 
                    <div></div>
                    <QuoteList quotes={quotes}/>
                    <div></div>
                    </main>
                  )}
            
    </div>
  )
}
