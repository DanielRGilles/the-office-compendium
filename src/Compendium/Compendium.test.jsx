import { render } from '@testing-library/react'

import Compendium from './Compendium'

it('Should fetch then render a list of office quotes', async () => {
  const { container } = render(<Compendium />)
  expect(container).toMatchSnapshot()
})
