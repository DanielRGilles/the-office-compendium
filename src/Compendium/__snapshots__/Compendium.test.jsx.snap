// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should fetch then render a list of office quotes 1`] = `
<div>
  <div
    class="main-body"
  >
    <h1
      class="header-box"
    >
      The Office
    </h1>
    <div>
      <form
        aria-label="searchForm"
      >
        <h3>
          Search for a word or character!
        </h3>
        <label
          for="word-search"
        >
          Query:
        </label>
        <input
          id="word-search"
          name="word"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Search!
        </button>
      </form>
      <div>
        <select>
          <option
            value="all"
          >
            All
          </option>
          <option
            value="Michael"
          >
            Michael
          </option>
          <option
            value="Pam"
          >
            Pam
          </option>
          <option
            value="Jim"
          >
            Jim
          </option>
          <option
            value="Andy"
          >
            Andy
          </option>
          <option
            value="Angela"
          >
            Angela
          </option>
          <option
            value="Charles"
          >
            Charles
          </option>
          <option
            value="Creed"
          >
            Creed
          </option>
          <option
            value="Darryl"
          >
            Darryl
          </option>
          <option
            value="David"
          >
            David
          </option>
          <option
            value="Deangelo"
          >
            Deangelo
          </option>
          <option
            value="Dwight"
          >
            Dwight
          </option>
          <option
            value="Ed"
          >
            Ed
          </option>
          <option
            value="Erin"
          >
            Erin
          </option>
          <option
            value="Gabe"
          >
            Gabe
          </option>
          <option
            value="Holly"
          >
            Holly
          </option>
          <option
            value="Jan"
          >
            Jan
          </option>
          <option
            value="Jo"
          >
            Jo
          </option>
          <option
            value="Josh"
          >
            Josh
          </option>
          <option
            value="Kelly"
          >
            Kelly
          </option>
          <option
            value="Kevin"
          >
            Kevin
          </option>
          <option
            value="Oscar"
          >
            Oscar
          </option>
          <option
            value="Phyllis"
          >
            Phyllis
          </option>
          <option
            value="Ryan"
          >
            Ryan
          </option>
          <option
            value="Stanley"
          >
            Stanley
          </option>
          <option
            value="Toby"
          >
            Toby
          </option>
          <option
            value="Todd"
          >
            Todd
          </option>
        </select>
      </div>
      <form>
        <select>
          <option
            value="1"
          >
            1
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="10"
          >
            10
          </option>
          <option
            value="25"
          >
            25
          </option>
        </select>
        <button
          type="submit"
        >
          Get Random Quotes
        </button>
      </form>
    </div>
    <h1
      class="load-spin"
    >
      Loading...
    </h1>
  </div>
</div>
`;
